onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorMatMulNBits.cpp

```diff
@@ -0,0 +1,195 @@
+        scaleToDequantizeEdge.GraphInputIndex = 2;
+        scaleToDequantizeEdge.ToNodeIndex = 0;
+        scaleToDequantizeEdge.ToNodeInputIndex = 1;
+        inputEdges.push_back(scaleToDequantizeEdge);
+
+        if (hasZeroPoint)
+        {
+            DML_INPUT_GRAPH_EDGE_DESC zeroPointToDequantizeEdge = {};
+            zeroPointToDequantizeEdge.GraphInputIndex = 3;
+            zeroPointToDequantizeEdge.ToNodeIndex = 0;
+            zeroPointToDequantizeEdge.ToNodeInputIndex = 2;
+            inputEdges.push_back(zeroPointToDequantizeEdge);
+        }
+
+        DML_INPUT_GRAPH_EDGE_DESC firstInputToGemmEdge = {};
+        firstInputToGemmEdge.GraphInputIndex = 0;
+        firstInputToGemmEdge.ToNodeIndex = 1;
+        firstInputToGemmEdge.ToNodeInputIndex = 0;
+        inputEdges.push_back(firstInputToGemmEdge);
+
+        DML_INTERMEDIATE_GRAPH_EDGE_DESC dequantizeToGemmEdge = {};
+        dequantizeToGemmEdge.FromNodeIndex = 0;
+        dequantizeToGemmEdge.FromNodeOutputIndex = 0;
+        dequantizeToGemmEdge.ToNodeIndex = 1;
+        dequantizeToGemmEdge.ToNodeInputIndex = 1;
+        intermediateEdges.push_back(dequantizeToGemmEdge);
+
+        DML_OUTPUT_GRAPH_EDGE_DESC gemmToOutputEdge = {};
+        gemmToOutputEdge.FromNodeIndex = 1;
+        gemmToOutputEdge.FromNodeOutputIndex = 0;
+        gemmToOutputEdge.GraphOutputIndex = 0;
+        outputEdges.push_back(gemmToOutputEdge);
+
+        MLOperatorGraphDesc operatorGraphDesc = {};
+        operatorGraphDesc.inputEdgeCount = gsl::narrow_cast<uint32_t>(inputEdges.size());
+        operatorGraphDesc.inputEdges = inputEdges.data();
+        operatorGraphDesc.intermediateEdgeCount = gsl::narrow_cast<uint32_t>(intermediateEdges.size());
+        operatorGraphDesc.intermediateEdges = intermediateEdges.data();
+        operatorGraphDesc.outputEdgeCount = gsl::narrow_cast<uint32_t>(outputEdges.size());
+        operatorGraphDesc.outputEdges = outputEdges.data();
+        operatorGraphDesc.nodeCount = gsl::narrow_cast<uint32_t>(opDescs.size());
+        operatorGraphDesc.nodes = opDescs.data();
+
+        SetDmlOperatorGraphDesc(std::move(operatorGraphDesc), kernelInfo);
+    }
+};
+
+void CALLBACK QueryMatMulNBits(IMLOperatorSupportQueryContextPrivate* context, /*out*/ bool* isSupported)
+{
+    *isSupported = false;
+
+    MLOperatorAttributes attributes(context);
+    const auto bitCount = attributes.GetAttribute<int64_t>(AttrName::Bits);
+    if (bitCount != 4 && bitCount != 8)
+    {
+        return;
+    }
+
+    // DML only supports perfect blocks at the moment, which is what most quantization tools produce anyways since they pad
+    // as part of the quantization process
+    const uint32_t bColCount = gsl::narrow_cast<uint32_t>(attributes.GetAttribute<int64_t>(AttrName::UppercaseK));
+    const auto blockSize = attributes.GetAttribute<int64_t>(AttrName::MatMulNBitsBlockSize);
+    if (bColCount % blockSize != 0)
```

Comment:
Should sanity check `blockSize` > 0 before potentially dividing by zero.

```suggestion
    if (blockSize <= 0 || bColCount % blockSize != 0)
```

or
`ML_CHECK_VALID_ARGUMENT(blockSize > 0);`